core/eth/state.go

```diff
@@ -59,7 +85,131 @@ func (cs *ChainState) GetOperatorSocket(ctx context.Context, blockNumber uint, o
 	return socket, nil
 }
 
-func getOperatorState(operatorsByQuorum core.OperatorStakes, blockNumber uint32) (*core.OperatorState, error) {
+// buildSocketMap returns a map from operatorID to socket address for the operators in the operatorsByQuorum
+func (cs *ChainState) buildSocketMap(ctx context.Context, operatorIds []core.OperatorID) error {
+	socketMap := make(map[core.OperatorID]*string)
+	for _, operatorID := range operatorIds {
+		// if the socket is already in the map, skip
+		if _, ok := socketMap[operatorID]; ok {
+			continue
+		}
+		socket, err := cs.Tx.GetOperatorSocket(ctx, operatorID)
+		if err != nil {
+			return err
+		}
+		socketMap[operatorID] = &socket
+	}
+
+	cs.socketMu.Lock()
+	for operatorID, socket := range socketMap {
+		cs.SocketMap[operatorID] = socket
+	}
+	cs.socketMu.Unlock()
+
+	return nil
+}
+
+// indexSocketMap indexes a block range of socket update events
+func (cs *ChainState) indexSocketMap(ctx context.Context) error {
+	currentBlockNumber, err := cs.GetCurrentBlockNumber()
+	if err != nil {
+		return err
+	}
+
+	registryCoordinator, err := cs.Tx.RegistryCoordinator(ctx)
+	if err != nil {
+		return err
+	}
+
+	logs, err := cs.Client.FilterLogs(ctx, ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(cs.socketPrevBlockNumber)),
+		ToBlock:   big.NewInt(int64(currentBlockNumber)),
+		Addresses: []gcommon.Address{registryCoordinator},
+		Topics: [][]gcommon.Hash{
+			{common.OperatorSocketUpdateEventSigHash},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(logs) == 0 {
+		return nil
+	}
+
+	var socketUpdates []*socketUpdateParams
+
+	// logs are in order of block number, so we can just iterate through them
+	for _, log := range logs {
+		tx, isPending, err := cs.Client.TransactionByHash(ctx, log.TxHash)
+		if err != nil {
+			return err
+		}
+		if isPending {
+			return fmt.Errorf("transaction pending for operator socket update event")
```

Comment:
I think this could be ok for the first implementation, but I think the following improvement should be made:

Currently, you are erroring out if *any* socket update transaction is still pending. So, if there were 100 updates, and the very last one is still pending, you would discard all the work that's already been done for the first 99 updates. This could be problematic if there are is a stead stream of new updates, which prevent older updates from being processed.

I think a better solution would be to process the updates one at a time, so that when you reach an update which is still pending, you can return from the method, and try again next time.

Reply:
good point! I took the parsing out to process one log at a time